### PR TITLE
Add ICAL Download Support

### PR DESCRIPTION
**Pull request by JishnuT09**
This pull request adds a new feature to allow users to download their schedule as an iCal calendar file. The main changes include adding a "Download Calendar" button to the UI, implementing logic to extract schedule data from the table, and generating a valid iCal file for download.

### CODE REVIEW

File: contributors.html [hunk 1]
```diff
@@ -28,6 +28,8 @@ <h1>Contributors</h1>
     </li>
     <li>Anmol Singh &lt;<a href="mailto:asingh1@imsa.edu">asingh1@imsa.edu</a>&gt;
     </li>
+    <li>Jishnu Thakkar <<a href="jthakkar@imsa.edu">jthakkar@imsa.edu</a>>
```
Comment (geosir) [inline contributors.html]: `<` and `>` are reserved HTML characters. Please use the HTML escape characters `&lt;` to show a `<` and `&gt;` to show a `>` (see rows above).

File: index.html [hunk 2]
```diff
@@ -289,6 +290,12 @@ <h2>Your Schedule</h2>
                 onClick="printDiv()">
             Print
         </button>
+        <button id="save-button"
```
Comment (geosir) [inline index.html:292]: Could you please add `#save-button` to the style in scheduler.css used by the other buttons? (Or better yet, modify that style to apply to those buttons using a class, since it's so common?)

File: scheduler.js [hunk 3]
```diff
@@ -399,3 +409,166 @@ function invert(obj) {
     }
     return new_obj;
 }
+
+// Extract schedule data from the table for iCal conversion
+getScheduleData = function () {
```
Comment (geosir) [inline scheduler.js]: Nice work reverse-engineering the schedule table!

For maintainability, I'd like an alternative to relying on the table structure (directly reading the `<b>`, `<u>` and `<p>` tags), since this could break if someone updates the structure of the table in the future. Two suggestions off the top of my head:

1. Class data is stored in the `data-schedule` attribute of each cell, and could be extracted from there instead.

2. (Better Design) The new `scheduleData` data structure makes a lot of sense, adding a central place to store a representation of the schedule and making it easier for future contributors to work with the data. First, consider changing `scheduleData` from an array into an object (dictionary) so that classes can be looked up efficiently if they need to be modified/deleted. Then, you could add logic to the `addToTable()` and `removeFromTable()` functions that directly updates `scheduleData` from the input. Then, when `downloadIcal()` is called, `scheduleData` will already be the latest data, and there will be no need to extract the schedule again.

File: scheduler.js [hunk 4]
```diff
@@ -5,6 +5,17 @@ function getParameterByName(name) {
     return urlParams.get(name);
 }
 
+class ScheduleItem {
+    constructor(time, className, room, teacher, daysOfWeek, durationMinutes) {
+        this.time = time;
```
Comment (geosir) [inline scheduler.js]: Although classes might (currently?) be the same mod each day, it's a little confusing that it is possible to make a schedule like the one below, but the iCal invite creates events with times all equal to the first class (8am). Could this class definition be modified to represent class times as mod/day pairs? Then, the iCal code could create an event for each mod/day pair.

(This might be easiest by modifying `addToTable` as mentioned below, since it already works with mod/day pairs for each class instance.)

image

File: scheduler.js [hunk 5]
```diff
@@ -399,3 +409,166 @@ function invert(obj) {
...
+                }
+            }
+        });
+    });
+    return scheduleData;
+}
+
+// Download the iCal file
+function downloadIcal() {
+    /*let email = document.getElementById("email").value;
```
Comment (geosir) [inline scheduler.js]: Please remove this comment if it's unused.

File: index.html [hunk 6]
```diff
@@ -8,6 +8,7 @@
     <link rel="stylesheet" href="scheduler.css">
 
     <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
+    <script src="https://apis.google.com/js/api.js"></script>
```
Comment (geosir) [inline index.html]: Where is this import used?

**Review by geosir — CHANGES_REQUESTED** [review]
Hi JishnuT09, thank you for the contribution - this is a really convenient feature!

I've just added some feedback and design suggestions. Once those have been addressed, this will be good to merge.

Please let me know if you have any questions. Thanks again!

**Comment by JishnuT09** [comment]
@geosir Thanks so much for the feedback. I went through the code and believe I implemented all of your changes. I also added a small check that doesn't allow for multiple classes to randomly be assigned the same colors. Sorry for the long delay in my response, and let me know if there's any more improvements I can make.
